caas/kubernetes/provider/metadata_test.go

```diff
@@ -170,344 +171,1134 @@ var hostRegionsTestCases = []hostRegionTestcase{
+				},
+			},
+		},
+		{
+			Name: "Test microk8s cloud prefers nominated storage as first priority",
+			InitialObjects: []runtime.Object{
+				newNode(map[string]string{
+					"failure-domain.beta.kubernetes.io/region": "wallyworld-region",
+					"microk8s.io/cluster":                      "true",
+				}),
+				&storagev1.StorageClass{
+					ObjectMeta: v1.ObjectMeta{
+						Name: "mynode",
+					},
+					Provisioner: "microk8s.io/hostpath",
+				},
+				&storagev1.StorageClass{
+					ObjectMeta: v1.ObjectMeta{
+						Name: "annotation1",
+						Annotations: map[string]string{
+							"juju.is/operator-storage": "true",
+						},
+					},
+				},
+				&storagev1.StorageClass{
+					ObjectMeta: v1.ObjectMeta{
+						Name: "annotation2",
+						Annotations: map[string]string{
+							"juju.is/workload-storage": "true",
+						},
+					},
+				},
+				&storagev1.StorageClass{
+					ObjectMeta: v1.ObjectMeta{
+						Name: "nominated",
+					},
+				},
+			},
+			NominatedStorage: "nominated",
+			Result: kubernetes.ClusterMetadata{
+				Cloud:   "microk8s",
+				Regions: set.NewStrings("wallyworld-region"),
+				WorkloadStorageClass: &storagev1.StorageClass{
+					ObjectMeta: v1.ObjectMeta{
+						Name: "nominated",
+					},
+				},
+				OperatorStorageClass: &storagev1.StorageClass{
+					ObjectMeta: v1.ObjectMeta{
+						Name: "nominated",
+					},
+				},
+			},
+		},
+		{
+			Name: "Test microk8s cloud with no found storage",
+			InitialObjects: []runtime.Object{
+				newNode(map[string]string{
+					"failure-domain.beta.kubernetes.io/region": "wallyworld-region",
+					"microk8s.io/cluster":                      "true",
+				}),
+			},
+			Result: kubernetes.ClusterMetadata{
+				Cloud:                "microk8s",
+				Regions:              set.NewStrings("wallyworld-region"),
+				WorkloadStorageClass: nil,
+				OperatorStorageClass: nil,
+			},
+		},
+		{
+			Name: "Test microk8s cloud doesn't use default storage",
+			InitialObjects: []runtime.Object{
+				newNode(map[string]string{
+					"failure-domain.beta.kubernetes.io/region": "wallyworld-region",
+					"microk8s.io/cluster":                      "true",
+				}),
+				&storagev1.StorageClass{
+					ObjectMeta: v1.ObjectMeta{
+						Name: "default",
+						Annotations: map[string]string{
+							"storageclass.kubernetes.io/is-default-class": "true",
+						},
+					},
+				},
+			},
+			Result: kubernetes.ClusterMetadata{
+				Cloud:                "microk8s",
+				Regions:              set.NewStrings("wallyworld-region"),
+				WorkloadStorageClass: nil,
+				OperatorStorageClass: nil,
+			},
+		},
 
-func (s *K8sMetadataSuite) TestMergeSelectors(c *gc.C) {
-	selector1 := k8slabels.SelectorFromSet(map[string]string{"foo": "bar"})
-	selector2 := k8slabels.SelectorFromSet(map[string]string{"foo1": "bar1"})
-	c.Assert(provider.MergeSelectors(selector1, selector2), gc.DeepEquals,
-		k8slabels.SelectorFromSet(map[string]string{
-			"foo":  "bar",
-			"foo1": "bar1",
-		}),
-	)
+		// Azure
```

Comment:
The same set of repeated tests for each cloud type - I reckon we can parameterise a common set of tests using cloud specific annotations, provisioning class etc as input. I tend to be allergic to lots of boilerplate / cut and paste code.

Reply:
It makes perfect sense to think that but each cloud is unique in their rules and while they may be very common today tomorrow they may change and if we are trying a one size fits all pattern then something may break and our signal will not catch it.

Reply:
It was the common annotation lookup stuff I think that could be consolidated.